Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternConstructorStringParser.h"
+
+#include "URLPatternComponent.h"
+#include "URLPatternInit.h"
+#include "URLPatternTokenizer.h"
+
+namespace WebCore {
+using namespace JSC;
+
+URLPatternConstructorStringParser::URLPatternConstructorStringParser(String&& input)
+    : m_input(WTFMove(input))
+{
+}
+
+// https://urlpattern.spec.whatwg.org/#rewind
+void URLPatternConstructorStringParser::rewind()
+{
+    m_tokenIndex = m_componentStart;
+    m_tokenIncrement = 0;
+}
+
+// https://urlpattern.spec.whatwg.org/#get-a-safe-token
+const URLPatternUtilities::Token& URLPatternConstructorStringParser::getSafeToken(size_t index) const
+{
+    if (index < m_tokenList.size())
+        return m_tokenList[index];
+
+    ASSERT(m_tokenList.size() >= 1);
+    ASSERT(m_tokenList.last().type == URLPatternUtilities::TokenType::End);
+    return m_tokenList.last();
+}
+
+// https://urlpattern.spec.whatwg.org/#is-a-non-special-pattern-char
+bool URLPatternConstructorStringParser::isNonSpecialPatternChar(size_t index, char value) const
+{
+    auto token = getSafeToken(index);
+    // FIXME: Find a better way to convert a char to a string.
+    StringView fixme = StringView(std::span<const char>(&value, 1));
```

Comment:
maybe:
```
token.value.length() == 1 && token.value[0] == value
```

Reply:
Alternatively, we can update isNonSpecialPatternChar() to take an `ASCIILiteral` instead of a `char`, assert that the length of the ASCIILiteral is 1 and simply use `token.value == value`.

Reply:
Oh cool I was just about to tag you two! Thanks I'll do the first one ha.